### PR TITLE
Fix text field label animation duration and curve

### LINKED ISSUE #18066: Text label animations are too slow compared to Material Design guidelines
The animations when a text label is being activated is too slow. In the guidelines the animations are at least twice the speed.

https://material.io/design/components/text-fields.html#anatomy

```
[✓] Flutter (Channel master, v0.5.2-pre.18, on Mac OS X 10.13.4 17E202, locale en-US)
    • Flutter version 0.5.2-pre.18 at /Users/vik/Projects/flutter
    • Framework revision 758711c85d (4 hours ago), 2018-05-31 10:33:15 +0200
    • Engine revision d174c4ff01
    • Dart version 2.0.0-dev.59.0.flutter-ff815d05a5

[!] Android toolchain - develop for Android devices (Android SDK 27.0.3)
    • Android SDK at /Users/vik/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-27, build-tools 27.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-915-b08)
    ✗ Android license status unknown.

[✓] iOS toolchain - develop for iOS devices (Xcode 9.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 9.3.1, Build version 9E501
    • ios-deploy 1.9.2
    • CocoaPods version 1.5.3

[✓] Android Studio (version 3.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    • Dart plugin version 171.4424
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-915-b08)

[✓] Android Studio (version 2.3)
    • Android Studio at /Applications/Android Studio - old.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] IntelliJ IDEA Community Edition (version 2017.2.5)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 20.0.2
    • Dart plugin version 172.4343.25

[✓] Connected devices (1 available)
    • iPhone 6 • 5DE7BCCC-8793-413A-9924-3D413DE12FCD • ios • iOS 9.0 (simulator)

! Doctor found issues in 1 category.
```

### PR DESCRIPTION
**Pull request by pourqavam**
Fixes https://github.com/flutter/flutter/issues/18066

This PR increases speed of text field label animation and add the missing curve to it.
The 167ms comes from here: [Android Material Text Field](https://github.com/material-components/material-components-android/blob/master/lib/java/com/google/android/material/textfield/TextInputLayout.java#L194)

Other references:
- https://material.io/components/text-fields#interactive-demo
- https://material.io/design/motion/speed.html#duration

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I signed the [CLA].
- [X] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by Pourqavam** [comment]
> @Pourqavam thanks for contributing! Could you add a test to verify the correct behavior? You might be able to find examples here: https://github.com/flutter/flutter/blob/master/packages/flutter/test/material/input_decorator_test.dart

New test added, old tests updated.

File: packages/flutter/lib/src/material/input_decorator.dart [hunk 1]
```diff
@@ -1862,6 +1863,7 @@ class InputDecorator extends StatefulWidget {
 
 class _InputDecoratorState extends State<InputDecorator> with TickerProviderStateMixin {
   late AnimationController _floatingLabelController;
+  late Animation<double> _floatingLabelAnimation;
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/material/input_decorator.dart:1866]: nit: `late final`?

File: packages/flutter/lib/src/material/input_decorator.dart [hunk 2]
```diff
@@ -16,7 +16,7 @@ import 'material_state.dart';
 import 'theme.dart';
 import 'theme_data.dart';
 
-const Duration _kTransitionDuration = Duration(milliseconds: 200);
+const Duration _kTransitionDuration = Duration(milliseconds: 167);
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/material/input_decorator.dart:19]: nit: could you add a comment to explain the provenance of this value? Something like "value extracted from <link>"

File: packages/flutter/lib/src/material/input_decorator.dart [hunk 3]
```diff
@@ -1879,6 +1881,11 @@ class _InputDecoratorState extends State<InputDecorator> with TickerProviderStat
       value: labelIsInitiallyFloating ? 1.0 : 0.0,
     );
     _floatingLabelController.addListener(_handleChange);
+    _floatingLabelAnimation = CurvedAnimation(
+      parent: _floatingLabelController,
+      curve: _kTransitionCurve,
+      reverseCurve: _kTransitionCurve.flipped,
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/material/input_decorator.dart:1887]: I'm not super familiar with Android, but from https://github.com/material-components/material-components-android/blob/master/lib/java/com/google/android/material/textfield/TextInputLayout.java#L4150-L4168, it seems the curve is always `AnimationUtils.FAST_OUT_SLOW_IN_INTERPOLATOR`?

Same question for the change made to `_borderAnimation`.

File: packages/flutter/lib/src/widgets/text_selection.dart [hunk 4]
```diff
@@ -2015,6 +2015,7 @@ class _TextSelectionGestureDetectorState extends State<TextSelectionGestureDetec
     if (!_isDoubleTap) {
       widget.onSingleTapUp?.call(details);
       _lastTapOffset = details.globalPosition;
+      _doubleTapTimer?.cancel();
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/widgets/text_selection.dart:2018]: The change here doesn't seem to be related?

File: packages/flutter/lib/src/material/input_decorator.dart (hunk 3 above)
Comment (Pourqavam) [reply packages/flutter/lib/src/material/input_decorator.dart:1889]: If the flipped curve is not used for the reverse, the curve actually reversed on unfocus, something like slowOutFastIn.
I don't know it's normal behavior or not, but this is what actually happens and can be seen visually in slow motion.

File: packages/flutter/lib/src/widgets/text_selection.dart (hunk 4 above)
Comment (Pourqavam) [reply packages/flutter/lib/src/widgets/text_selection.dart:2018]: Without canceling the old timer, the `A Focused text-field will not lose focus when clicking on its decoration` test in the `text_field_focus_test.dart` will be failed with this error:

> A Timer is still pending even after the widget tree was disposed

Due to reducing the `_kTransitionDuration`, something happens too quickly.

**Review by LongCatIsLooong — APPROVED** [review]
LGTM. Could you sync this branch to `head`? Looks like there're merge conflicts.

**Comment by CaseyHillers** [comment]
FYI this was reverted in https://github.com/flutter/flutter/pull/114646 due to some race conditions in internal tests. @LongCatIsLooong can you help with the reland?

**Pull request merged**